### PR TITLE
feat: Add sink::unfold

### PR DESCRIPTION
**Pull request by Marwes**
A more general function than `from_fn` (#2254), as suggested in
https://github.com/rust-lang/futures-rs/pull/2254#issuecomment-731860678

### CODE REVIEW

**Review by taiki-e — APPROVED** [review]
Thanks, this looks good to me. I think we'll deprecate this and add something like #2254 once async closures became working properly, but for now, this is more flexible and I prefer this than #2254. (AFAIK, no one is currently working on async closures, so I don't think async closures' problems will be fixed soon.)

cc @cramertj @Nemo157: Any opinion on adding this? (I tend to merge this)

File: futures-util/src/sink/unfold.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+/// Sink for the [`unfold`] function.
+#[pin_project]
+#[derive(Debug)]
+#[must_use = "sinks do nothing unless polled"]
+pub struct Unfold<T, F, R> {
+    state: Option<T>,
+    function: F,
+    #[pin]
+    future: Option<R>,
```
Comment (piegamesde) [inline futures-util/src/sink/unfold.rs:14]: `state` and `future` are never present at the same time, so maybe the use of an enum could reduce the size of the struct?

Comment (taiki-e) [reply futures-util/src/sink/unfold.rs:14]: I would accept a patch to reduce the size of `unfold`.

**Comment by sdroege** [comment]
The way how this is implemented is slightly suboptimal for type inference, see https://github.com/rust-lang/futures-rs/pull/2311

**Pull request merged**